Sources/TensorFlow/Operators/LinearAlgebra.swift

```diff
@@ -266,3 +266,121 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    let broadcastMatrix: Bool = matrix.rank < rhs.rank
+    let broadcastRhs: Bool = matrix.rank > rhs.rank
+
+    var matchedMatrix: Tensor<T> = matrix
+    if broadcastMatrix {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        matchedMatrix = matrix.broadcasted(to: leadingDims + matrix.shape)
+    }
+    
+    let x = triangularSolve(matrix: matchedMatrix, rhs: rhs, lower: lower, adjoint: adjoint)
+    let triangularSolveGrad = { (gradient: Tensor<T>) -> (Tensor<T>, Tensor<T>) in
+        let rhsGrad = triangularSolve(matrix: matchedMatrix, rhs: gradient, lower: lower, adjoint: !adjoint)
+        let (left, right) = adjoint ? (x, rhsGrad) : (rhsGrad, x)
+        let matrixGrad = -matmul(left, transposed: false, right, transposed: true)
+        let triMatrixGrad = lower
+            ? matrixGrad.bandPart(subdiagonalCount: -1, superdiagonalCount: 0)
+            : matrixGrad.bandPart(subdiagonalCount: 0, superdiagonalCount: -1)
+        if broadcastMatrix {
+            return (triMatrixGrad.unbroadcasted(to: matrix.shape), rhsGrad)
+        } else if broadcastRhs {
+            return (triMatrixGrad, rhsGrad.unbroadcasted(to: rhs.shape))
+        }
+        return (triMatrixGrad, rhsGrad)
+    }
+    return (x, triangularSolveGrad)
+}
+
+
+/// Returns leading dimension from tensors shapes.
+/// Example:
+/// The leading dimension for tensors with shapes `[N, M, K]` and `[M, K]`
+/// is `[N]`. In cases, when tensors shapes do not have common parts from the end, e.g.
+/// `[N, M, K]` and `[N, M]`, it will raise an error.
+///
+/// - Parameters:
+///     - left: Left tensor.
+///     - right: Right tensor.
+///     - ignoreLast: Number of dimensions from the end to ignore. Default value is zero.
+/// - Precondition: Left tensor rank cannot be lower than `ignoreLast`.
+/// - Precondition: Right tensor rank cannot be lower than `ignoreLast`.
+@inlinable
+public func extractLeadingDims<T: TensorFlowNumeric>(
+    _ left: Tensor<T>,
+    _ right: Tensor<T>,
+    ignoreLast: Int = 0
+) -> TensorShape {
+    extractLeadingDims(left.shape, right.shape, ignoreLast: ignoreLast)
+}
+
+/// Returns leading dimension from tensors shapes.
+/// Example:
+/// The leading dimension for tensors with shapes `[N, M, K]` and `[M, K]`
+/// is `[N]`. In cases, when tensors shapes do not have common parts from the end, e.g.
+/// `[N, M, K]` and `[N, M]`, it will raise an error.
+///
+/// - Parameters:
+///     - left: Left tensor shape.
+///     - right: Right tensor shape.
+///     - ignoreLast: Number of dimensions from the end to ignore. Default value is zero.
+/// - Precondition: Left tensor rank cannot be lower than `ignoreLast`.
+/// - Precondition: Right tensor rank cannot be lower than `ignoreLast`.
+@inlinable
+public func extractLeadingDims(_ left: TensorShape, _ right: TensorShape, ignoreLast: Int = 0) -> TensorShape {
+    precondition(left.rank >= ignoreLast, "The left tensor must have at least rank 2.")
```

Comment:
```suggestion
    precondition(left.rank >= ignoreLast, "The left tensor must have at least rank `ignoreLast`.")
```
Same for the `right.rank` precondition.

Reply:
The `right.rank` precondition wasn't updated.

Reply:
done